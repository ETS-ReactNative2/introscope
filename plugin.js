const toPairs = obj => {
    const pairs = []
    for (const key in obj) {
        pairs.push([key, obj[key]])
    }
    return pairs
}

const byType = type => node => node.type == type
const not = fn => (...args) => !fn(...args)

function processProgram({ types: t }, programPath, programState) {
    let options = { removeImports: false }

    const scopeId = programPath.scope.generateUidIdentifier('scope')

    const scopeSet = (left, right) =>
        t.assignmentExpression('=', t.memberExpression(scopeId, left), right)

    const unwrapOrRemoveExports = path => {
        path.get('body').forEach(path => {
            if (path.isExportDefaultDeclaration()) {
                path.replaceWith(
                    t.expressionStatement(
                        scopeSet(
                            t.identifier('default'),
                            path.get('declaration').node
                        )
                    )
                )
            } else if (path.isExportNamedDeclaration()) {
                path.node.declaration
                    ? path.replaceWith(path.get('declaration').node)
                    : path.remove()
            } else if (path.isExportAllDeclaration()) {
                path.remove()
            }
        })
    }

    const identifiersToObjectProperties = identifiers =>
        identifiers.map(identifier =>
            t.objectProperty(identifier, identifier, undefined, true)
        )

    const wrapInFunction = (globalIds, body) =>
        t.functionExpression(
            null,
            [t.assignmentPattern(scopeId, t.objectExpression([]))],
            t.blockStatement(
                []
                    .concat(
                        globalIds.length
                            ? t.variableDeclaration(
                                  'var',
                                  globalIds.map(id => t.variableDeclarator(id))
                              )
                            : []
                    )
                    .concat(body)
                    .concat([t.returnStatement(scopeId)])
            )
        )

    const getGlobalIdentifiers = scope =>
        toPairs(scope.globals).map(([_, identifier]) => identifier)

    const moduleExports = right =>
        t.expressionStatement(
            t.assignmentExpression(
                '=',
                t.memberExpression(
                    t.identifier('module'),
                    t.identifier('exports')
                ),
                right
            )
        )

    const variableDeclaratorToScope = (path, identifier) => {
        const idPath = path.get('id')
        if (idPath.isObjectPattern()) {
            path.insertAfter(
                t.variableDeclarator(
                    path.scope.generateUidIdentifier('temp'),
                    scopeSet(t.clone(identifier), t.clone(identifier))
                )
            )
        } else if (idPath.isArrayPattern()) {
        } else {
            const init = path.get('init')
            if (!init.node) return
            // Q: Why not replace the parent VariableDeclaration node?
            // A: To not die debugging variable declaration order in a `for` loop if one of the binding should be ignored dew to introscope configuration (expected future feature). And it's simplier to implement :)
            init.replaceWith(scopeSet(path.get('id').node, init.node))
        }
    }

    const classDeclarationToScope = path => {
        const classExpression = t.clone(path.node)
        classExpression.type = 'ClassExpression'
        path.replaceWith(
            t.expressionStatement(
                scopeSet(path.get('id').node, classExpression)
            )
        )
    }

    const functionDeclarationToScope = path => {
        const program = path.findParent(path => path.isProgram())
        program.unshiftContainer(
            'body',
            t.expressionStatement(
                scopeSet(path.get('id').node, path.get('id').node)
            )
        )
    }

    const declarationToScope = (path, identifier) => {
        if (path.isNodeType('VariableDeclarator')) {
            variableDeclaratorToScope(path, identifier)
        } else if (path.isNodeType('ClassDeclaration')) {
            classDeclarationToScope(path)
        } else if (path.isNodeType('FunctionDeclaration')) {
            functionDeclarationToScope(path)
        } else if (
            path.isNodeType('ImportDefaultSpecifier') ||
            path.isNodeType('ImportSpecifier') ||
            path.isNodeType('ImportNamespaceSpecifier')
        ) {
            if (options.removeImports === true) {
                // ignore
            } else if (path.parentPath.isImportDeclaration()) {
                return path.get('local').node
            }
        } else {
            throw new TypeError('Unknown node.type = ' + path.node.type)
            // TODO: log warning here using babel logger
        }
    }

    const replaceReferenceWithScope = path => {
        // ExportNamedDeclaration gets properly processed by replaceMutationWithScope()
        if (path.node.type == 'ExportNamedDeclaration') return
        // ExportSpecifier gets removed by unwrapOrRemoveExports()
        if (path.parent.type == 'ExportSpecifier') return
        if (path.findParent(path => path.isObjectPattern())) return

        const scoped = t.memberExpression(scopeId, path.node)
        if (path.parent && path.parent.type == 'CallExpression') {
            path.replaceWith(
                t.sequenceExpression([t.numericLiteral(0), scoped])
            )
        } else {
            path.replaceWith(scoped)
        }
    }

    const replaceMutationWithScope = path => {
        if (path.node.type == 'AssignmentExpression') {
            const left = path.get('left')
            left.replaceWith(t.memberExpression(scopeId, left.node))
        }
    }

    const bindingToScope = binding => {
        binding.referencePaths.forEach(replaceReferenceWithScope)
        binding.constantViolations.forEach(replaceMutationWithScope)
        return declarationToScope(binding.path, binding.identifier)
    }

    const bindingsToScope = bindings =>
        toPairs(bindings)
            .map(([_, binding]) => binding)
            .map(bindingToScope)
            .filter(Boolean)

    const program = (path, state) => {
        const globalIds = getGlobalIdentifiers(path.scope)
        const programGlobalNames = Object.keys(path.scope.globals)

        const localImportIds = bindingsToScope(path.scope.bindings)

        // unwrapOrRemoveExports() should go after bindingsToScope() as the latter treats `export var/let/const` as a reference and uses node.parent to distinguish
        unwrapOrRemoveExports(path)

        // reverse()-ing to preserve order after unshift()-ing
        localImportIds
            .reverse()
            .forEach(localId =>
                path.node.body.unshift(
                    t.expressionStatement(scopeSet(localId, localId))
                )
            )

        const programBody = path.node.body
        if (options.removeImports === true) {
            path.node.body = []
        } else {
            const importsOnly = programBody.filter(byType('ImportDeclaration'))
            path.node.body = importsOnly
        }
        const bodyWithoutImports = programBody.filter(
            not(byType('ImportDeclaration'))
        )

        path.pushContainer(
            'body',
            moduleExports(wrapInFunction(globalIds, bodyWithoutImports))
        )

        const uniqueGlobalIds = new Set()
        path.traverse({
            Scope(path, state) {
                programGlobalNames.forEach(globalName => {
                    const binding = path.scope.getBinding(globalName)
                    uniqueGlobalIds.add(binding)
                })
            }
        })
        bindingsToScope(Array.from(uniqueGlobalIds.values()))
    }

    program(programPath, programState)
}

export default function(babel) {
    return {
        visitor: {
            Program(path, state) {
                processProgram(babel, path, state)
            }
        }
    }
}
