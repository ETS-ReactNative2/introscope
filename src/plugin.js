const toPairs = obj => {
    const pairs = [];
    for (const key in obj) {
        pairs.push([key, obj[key]]);
    }
    return pairs;
};

const byType = type => node => node.type == type;
const not = fn => (...args) => !fn(...args);

function processProgram({ types: t }, programPath, programState) {
    const defaultOptions = {
        enable: false,
        ignore: [],
        removeImports: false
    };

    let options = {
        ...defaultOptions,
        ...programState.opts
    };

    const scopeId = programPath.scope.generateUidIdentifier('scope');

    const scopeSet = (left, right) =>
        t.assignmentExpression('=', t.memberExpression(scopeId, left), right);

    const unwrapOrRemoveExports = path => {
        path.get('body').forEach(path => {
            if (path.isExportDefaultDeclaration()) {
                path.replaceWith(
                    t.expressionStatement(
                        scopeSet(
                            t.identifier('default'),
                            path.get('declaration').node
                        )
                    )
                );
            } else if (path.isExportNamedDeclaration()) {
                path.node.declaration
                    ? path.replaceWith(path.get('declaration').node)
                    : path.remove();
            } else if (path.isExportAllDeclaration()) {
                path.remove();
            }
        });
    };

    const identifiersToObjectProperties = identifiers =>
        identifiers.map(identifier =>
            t.objectProperty(identifier, identifier, undefined, true)
        );

    const wrapInFunction = (globalIds, body) =>
        t.functionExpression(
            null,
            [t.assignmentPattern(scopeId, t.objectExpression([]))],
            t.blockStatement(
                []
                    .concat(
                        globalIds.length
                            ? t.variableDeclaration(
                                  'var',
                                  globalIds.map(id => t.variableDeclarator(id))
                              )
                            : []
                    )
                    .concat(body)
                    .concat([t.returnStatement(scopeId)])
            )
        );

    const moduleExports = right =>
        t.expressionStatement(
            t.assignmentExpression(
                '=',
                t.memberExpression(
                    t.identifier('module'),
                    t.identifier('exports')
                ),
                right
            )
        );

    const variableDeclaratorToScope = (path, identifier) => {
        const idPath = path.get('id');
        if (idPath.isObjectPattern() || idPath.isArrayPattern()) {
            path.insertAfter(
                t.variableDeclarator(
                    path.scope.generateUidIdentifier('temp'),
                    scopeSet(t.clone(identifier), t.clone(identifier))
                )
            );
        } else {
            const init = path.get('init');
            if (!init.node) return;
            // Q: Why not replace the parent VariableDeclaration node?
            // A: To not die debugging variable declaration order in a `for` loop if one of the binding should be ignored dew to introscope configuration (expected future feature). And it's simplier to implement :)
            init.replaceWith(scopeSet(path.get('id').node, init.node));
        }
    };

    const classDeclarationToScope = path => {
        const classExpression = t.clone(path.node);
        classExpression.type = 'ClassExpression';
        path.replaceWith(
            t.expressionStatement(
                scopeSet(path.get('id').node, classExpression)
            )
        );
    };

    const functionDeclarationToScope = path => {
        const program = path.findParent(path => path.isProgram());
        program.unshiftContainer(
            'body',
            t.expressionStatement(
                scopeSet(path.get('id').node, path.get('id').node)
            )
        );
    };

    const declarationToScope = (path, identifier) => {
        if (path.isNodeType('VariableDeclarator')) {
            variableDeclaratorToScope(path, identifier);
        } else if (path.isNodeType('ClassDeclaration')) {
            classDeclarationToScope(path);
        } else if (path.isNodeType('FunctionDeclaration')) {
            functionDeclarationToScope(path);
        } else if (
            path.isNodeType('ImportDefaultSpecifier') ||
            path.isNodeType('ImportSpecifier') ||
            path.isNodeType('ImportNamespaceSpecifier')
        ) {
            if (options.removeImports === true) {
                // ignore
            } else if (path.parentPath.isImportDeclaration()) {
                return path.get('local').node;
            }
        } else {
            throw path.buildCodeFrameError(
                'Cannot apply declarationToScope() to an anknown node.type: ' +
                    path.node.type
            );
        }
    };

    const replaceReferenceWithScope = path => {
        // ExportNamedDeclaration gets properly processed by replaceMutationWithScope()
        if (path.node.type == 'ExportNamedDeclaration') return;
        // ExportSpecifier gets removed by unwrapOrRemoveExports()
        if (path.parent.type == 'ExportSpecifier') return;
        if (path.findParent(path => path.isObjectPattern())) return;

        const scoped = t.memberExpression(scopeId, path.node);
        if (path.parent && path.parent.type == 'CallExpression') {
            path.replaceWith(
                t.sequenceExpression([t.numericLiteral(0), scoped])
            );
        } else {
            path.replaceWith(scoped);
        }
    };

    const replaceMutationWithScope = path => {
        if (path.node.type == 'AssignmentExpression') {
            const left = path.get('left');
            left.replaceWith(t.memberExpression(scopeId, left.node));
        }
    };

    const bindingToScope = binding => {
        binding.referencePaths.forEach(replaceReferenceWithScope);
        binding.constantViolations.forEach(replaceMutationWithScope);
        return declarationToScope(binding.path, binding.identifier);
    };

    const bindingsToScope = bindings =>
        toPairs(bindings)
            .filter(([name, _]) => !options.ignore.includes(name))
            .map(([_, binding]) => binding)
            .map(bindingToScope)
            .filter(Boolean);

    const parseConfig = path => {
        const firstStatement = path.get('body').forEach(statement => {
            const leading = statement.node.leadingComments || [];
            const trailing = statement.node.trailingComments || [];
            const comments = [].concat(leading).concat(trailing);
            comments.map(node => node.value).forEach(comment => {
                const [_, configJSON] = comment.split(/^\s*@introscope-config/);
                if (configJSON) {
                    try {
                        const config = JSON.parse(`{${configJSON}}`);
                        options = {
                            ...options,
                            ...config
                        };
                    } catch (ex) {
                        console.error(
                            'Error parsing Introscope config:',
                            comment
                        );
                    }
                }
            });
        });
    };

    const program = (path, state) => {
        parseConfig(path);
        if (!options.enable) return;

        const globalIds = toPairs(path.scope.globals)
            .filter(([name, _]) => !options.ignore.includes(name))
            .map(([_, identifier]) => identifier);
        const programGlobalNames = Object.keys(path.scope.globals);

        const localImportIds = bindingsToScope(path.scope.bindings);

        // unwrapOrRemoveExports() should go after bindingsToScope() as the latter treats `export var/let/const` as a reference and uses node.parent to distinguish
        unwrapOrRemoveExports(path);

        // reverse()-ing to preserve order after unshift()-ing
        localImportIds
            .reverse()
            .forEach(localId =>
                path.node.body.unshift(
                    t.expressionStatement(scopeSet(localId, localId))
                )
            );

        const programBody = path.node.body;
        if (options.removeImports === true) {
            path.node.body = [];
        } else {
            const importsOnly = programBody.filter(byType('ImportDeclaration'));
            path.node.body = importsOnly;
        }
        const bodyWithoutImports = programBody.filter(
            not(byType('ImportDeclaration'))
        );

        path.pushContainer(
            'body',
            moduleExports(wrapInFunction(globalIds, bodyWithoutImports))
        );

        // globals become locals after wrapInFunction()
        // TODO: make it more explicit in the code
        const globalBindings = {};
        path.traverse({
            Scope(path, state) {
                programGlobalNames.forEach(globalName => {
                    const binding = path.scope.getBinding(globalName);
                    globalBindings[globalName] = binding;
                });
            }
        });
        bindingsToScope(globalBindings);
    };

    program(programPath, programState);
}

module.exports = function(babel) {
    return {
        visitor: {
            Program(path, state) {
                if (state.opts.disable) return;

                if (
                    typeof process == 'object' &&
                    process.env.NODE_ENV != 'test'
                ) {
                    return;
                }

                processProgram(babel, path, state);
            }
        }
    };
};
