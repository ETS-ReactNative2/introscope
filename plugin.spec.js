import { transform } from 'babel-core'
import plugin from './plugin'

const shoot = code =>
    expect(
        transform(code, {
            plugins: [plugin]
        }).code
    ).toMatchSnapshot()

describe('plugin', () => {
    it('import', () => {
        shoot(`
            import 'some-module1'
            import defaultImport from 'some-module2'
            import { singleNamedImport } from 'some-module3'
            import { namedImport1, namedImport2 } from 'some-module4'
            import { originalImportName as localImportName } from 'some-module5'
            import * as namespaceImport from 'some-module6'

            [
                defaultImport,
                singleNamedImport,
                namedImport1,
                namedImport2,
                localImportName,
                namespaceImport
             ]
        `)
    })

    it('assignments', () => {
        shoot(`
            let x;
            x++;
            x = 1;
        `)
    })

    it('references', () => {
        shoot(`
            let x;
            x();
            !function(){ x++ };
            () => { return { x } };
            !{ x }
        `)
    })

    it('declarations', () => {
        shoot(`
            let noInitValue;
            for (var forVar = 0; forVar != 0; forVar++);
            let singleInit = 1;
            let doubleInit1 = 1, doubleInit2 = 2;
            var sameDoubleInit = 1, sameDoubleInit = 2;

            let { singleDestructing } = 1
            let { objectDestruction1, objectDestruction2, nestedProperty: { objectNestedDestruction1 } } = { objectDestruction1, objectDestruction2, nestedProperty: { objectNestedDestruction1 } }, let2 = objectDestruction1;
            // let { ...objectRest } = 1;
            let [ arrayDestruction1, arrayDestruction2, [ arrayNestedDestruction ] ] = 1;
            //let [ ...arrayRest ] ] = 1;


            class ClassName {}; new ClassName();
            functionName(); function functionName () {}
        `)
    })

    it('scopes', () => {
        shoot(`
            let x = true;
            for (let x = 0; x != 0; x++);
            { let x = false; }
            !function(){ var x = false; }
        `)
    })

    it('globals', () => {
        shoot(`
            let x = global1;
            global2 = 1;
            global2 = 1;
            globalFunction();
            !function(){ return [nestedGlobal1, global2] }();
        `)
    })

    it('export', () => {
        shoot(`
            let defaultExport, toBeNameExport1, toBeNameExport2;
            export default defaultExport;
            export const namedSingleExport = 'namedSingleExportValue'
            export { toBeNameExport1, toBeNameExport2 }
            export * from 'some-module'
        `)
    })
})

describe('options', () => {
    it('removeImport', () => {
        shoot(`
            // @introscope removeImports: ['defaultImport1', 'singleNamedImport1', 'namedImport1', 'namedImport2', 'localImportName1', 'namespaceImport1']
            import defaultImport1 from 'some-module1'
            import defaultImport2 from 'some-module2'
            import { singleNamedImport1 } from 'some-module2'
            import { singleNamedImport2 } from 'some-module3'
            import { namedImport1, namedImport2 } from 'some-module4'
            import { namedImport3, namedImport4 } from 'some-module5'
            import { originalImportName1 as localImportName1 } from 'some-module6'
            import { originalImportName2 as localImportName2 } from 'some-module7'
            import * as namespaceImport1 from 'some-module8'
            import * as namespaceImport2 from 'some-module9'

            [
                defaultImport1,
                defaultImport2,
                singleNamedImport1,
                singleNamedImport2,
                namedImport1,
                namedImport2,
                namedImport3,
                namedImport4,
                localImportName1,
                localImportName2,
                namespaceImport1,
                namespaceImport2
             ]
        `)
    })
})
