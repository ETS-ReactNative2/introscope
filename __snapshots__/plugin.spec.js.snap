// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options removeImport 1`] = `
"
// @introscope removeImports: ['defaultImport1', 'singleNamedImport1', 'namedImport1', 'namedImport2', 'localImportName1', 'namespaceImport1']
import defaultImport1 from 'some-module1';
import defaultImport2 from 'some-module2';
import { singleNamedImport1 } from 'some-module2';
import { singleNamedImport2 } from 'some-module3';
import { namedImport1, namedImport2 } from 'some-module4';
import { namedImport3, namedImport4 } from 'some-module5';
import { originalImportName1 as localImportName1 } from 'some-module6';
import { originalImportName2 as localImportName2 } from 'some-module7';
import * as namespaceImport1 from 'some-module8';
import * as namespaceImport2 from 'some-module9';

module.exports = function (_scope = {}) {
    _scope.defaultImport1 = defaultImport1;
    _scope.defaultImport2 = defaultImport2;
    _scope.singleNamedImport1 = singleNamedImport1;
    _scope.singleNamedImport2 = singleNamedImport2;
    _scope.namedImport1 = namedImport1;
    _scope.namedImport2 = namedImport2;
    _scope.namedImport3 = namedImport3;
    _scope.namedImport4 = namedImport4;
    _scope.localImportName1 = localImportName1;
    _scope.localImportName2 = localImportName2;
    _scope.namespaceImport1 = namespaceImport1;
    _scope.namespaceImport2 = namespaceImport2;
    [_scope.defaultImport1, _scope.defaultImport2, _scope.singleNamedImport1, _scope.singleNamedImport2, _scope.namedImport1, _scope.namedImport2, _scope.namedImport3, _scope.namedImport4, _scope.localImportName1, _scope.localImportName2, _scope.namespaceImport1, _scope.namespaceImport2];
    return _scope;
};"
`;

exports[`plugin assignments 1`] = `
"module.exports = function (_scope = {}) {
            let x;
            _scope.x++;
            _scope.x = 1;
            return _scope;
};"
`;

exports[`plugin declarations 1`] = `
"module.exports = function (_scope = {}) {
            _scope.functionName = functionName;

            let noInitValue;
            for (var forVar = _scope.forVar = 0; _scope.forVar != 0; _scope.forVar++);
            let singleInit = _scope.singleInit = 1;
            let doubleInit1 = _scope.doubleInit1 = 1,
                doubleInit2 = _scope.doubleInit2 = 2;
            var sameDoubleInit = 1,
                sameDoubleInit = _scope.sameDoubleInit = 2;
            _scope.ClassName = class ClassName {};
            ;new _scope.ClassName();
            (0, _scope.functionName)();function functionName() {}
            return _scope;
};"
`;

exports[`plugin export 1`] = `
"module.exports = function (_scope = {}) {
            let defaultExport, toBeNameExport1, toBeNameExport2;
            _scope.default = _scope.defaultExport;
            const namedSingleExport = _scope.namedSingleExport = 'namedSingleExportValue';
            return _scope;
};"
`;

exports[`plugin globals 1`] = `
"module.exports = function (_scope = {}) {
            var global2, global1, globalFunction, nestedGlobal1;

            let x = _scope.x = _scope.global1;
            _scope.global2 = 1;
            _scope.global2 = 1;
            (0, _scope.globalFunction)();
            !function () {
                        return [_scope.nestedGlobal1, _scope.global2];
            }();
            return _scope;
};"
`;

exports[`plugin import 1`] = `
"
import 'some-module1';
import defaultImport from 'some-module2';
import { singleNamedImport } from 'some-module3';
import { namedImport1, namedImport2 } from 'some-module4';
import { originalImportName as localImportName } from 'some-module5';
import * as namespaceImport from 'some-module6';

module.exports = function (_scope = {}) {
            _scope.defaultImport = defaultImport;
            _scope.singleNamedImport = singleNamedImport;
            _scope.namedImport1 = namedImport1;
            _scope.namedImport2 = namedImport2;
            _scope.localImportName = localImportName;
            _scope.namespaceImport = namespaceImport;
            [_scope.defaultImport, _scope.singleNamedImport, _scope.namedImport1, _scope.namedImport2, _scope.localImportName, _scope.namespaceImport];
            return _scope;
};"
`;

exports[`plugin references 1`] = `
"module.exports = function (_scope = {}) {
            let x;
            (0, _scope.x)();
            !function () {
                        _scope.x++;
            };
            () => {
                        return { x: _scope.x };
            };
            return _scope;
};"
`;

exports[`plugin scopes 1`] = `
"module.exports = function (_scope = {}) {
            let x = _scope.x = true;
            for (let x = 0; x != 0; x++);
            {
                        let x = false;
            }
            !function () {
                        var x = false;
            };
            return _scope;
};"
`;
